sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java

```diff
@@ -164,68 +171,85 @@ private void onChange(final DDlogCommand<DDlogRecord> command) {
-        protected MockResult visitQuery(final Query node, final String context) {
+        protected MockResult visitQuery(final Query node, final QueryContext context) {
             final QuerySpecification specification = (QuerySpecification) node.getQueryBody();
             return visitQuerySpecification(specification, context);
         }
 
         @Override
-        protected MockResult visitInsert(final Insert node, final String sql) {
+        protected MockResult visitInsert(final Insert node, final QueryContext context) {
             try {
                 // The assertions below encode assumption A2 (see javadoc for the DDlogJooqProvider class)
                 if (!(node.getQuery().getQueryBody() instanceof Values)) {
-                    throw new RuntimeException("Statement not supported: " + sql);
+                    throw new RuntimeException("Statement not supported: " + context.sql());
                 }
                 final Values values = (Values) node.getQuery().getQueryBody();
                 final String tableName = node.getTarget().toString();
                 final List<Field<?>> fields = tablesToFields.get(tableName.toUpperCase());
                 final int tableId = dDlogAPI.getTableId(ddlogRelationName(tableName));
+                int bindingIndex = 0;
                 for (final Expression row: values.getRows()) {
                     if (!(row instanceof Row)) {
-                        throw new RuntimeException("Statement not supported: " + sql);
+                        throw new RuntimeException("Statement not supported: " + context.sql());
                     }
                     final List<Expression> items = ((Row) row).getItems();
                     if (items.size() != fields.size()) {
-                        throw new RuntimeException(
-                                String.format("Incorrect row size for insertion into table %s: %s", tableName, sql));
+                        final String error = String.format("Incorrect row size for insertion into table %s. " +
+                                                   "Please specify all the table's fields in their declared order: %s",
+                                                    tableName, context.sql());
+                        throw new RuntimeException(error);
                     }
                     final DDlogRecord[] recordsArray = new DDlogRecord[items.size()];
-                    for (int i = 0; i < items.size(); i++) {
-                        recordsArray[i] = parseLiterals.process(items.get(i), fields.get(i).getDataType().nullable());
+                    if (context.hasBinding()) {
+                        // Is a statement with bind variables
+                        for (int i = 0; i < items.size(); i++) {
+                            final boolean isNullableField = fields.get(i).getDataType().nullable();
+                            final DDlogRecord record = toValue(fields.get(i), context.binding()[bindingIndex]);
```

Comment:
Should this counter be maintained in the context and provided by a context.nextBinding() method?

Reply:
That's a good idea. Every use of the binding seems to bump an index. So it's better to keep it behind the Context.

Reply:
Done.